### PR TITLE
fix(voice): ignore empty first chunk for speaking event

### LINKED ISSUE #569: fix(voice/ts): AgentSpeakingEvent.set() fires on an empty first audio chunk (adapter.runtime.ts)
Flagged in the #561 /review (non-blocking — no observed failure). The drain in `javascript/src/voice/adapter.runtime.ts` (~:318) calls `speakingEvent.set()` on the first `receiveAudio()` even when `first.data.length === 0` → can falsely signal "agent speaking" on a no-audio turn, which the interrupt-truncation path keys off. Guard: only `set()` when the first chunk has audio. Adjacent to the barge-in/truncation logic fixed in #561. Epic #370.

### PR DESCRIPTION
**Pull request by allin2**
## Why

An empty first audio chunk does not itself prove that the agent spoke, but the TypeScript voice runtime resolved `AgentSpeakingEvent` immediately. That could make interruption logic treat a silent turn as active speech, while simply gating on the first chunk could miss real audio that follows an empty prefix.

Closes #569

## What changed

- Resolve `AgentSpeakingEvent` and start playback timing only when the drain observes its first playable audio chunk, whether that is the initial chunk or a later one.
- Keep the event unset for a response containing only empty terminal chunks, and preserve the public API.
- Cover both empty-only and empty-prefix-then-audio paths with deterministic in-memory regression tests.

## Test plan

- Confirmed `does not set the speaking event when the first chunk is empty` failed before the runtime fix (`expected false`, received `true`).
- Confirmed `sets the speaking event when audio follows a leading empty chunk` failed before the review fix (`expected true`, received `false`).
- `pnpm exec vitest run src/voice/__tests__/interrupt-truncation.test.ts` — 12 passed.
- `pnpm exec vitest run src/voice/__tests__ src/voice/adapters/__tests__` — 45 files passed, 1 skipped; 436 tests passed, 1 skipped.
- `pnpm test:ci` — 89 files passed, 1 skipped; 1036 tests passed, 4 skipped.
- `pnpm lint:all`
- `pnpm typecheck:all`
- `pnpm build:all`
- `pnpm smoke:dist` — CJS and ESM distributions load successfully.

## How I can prove I was successful

No playable artifact — this is an internal voice-runtime state fix. Run the named regression tests in the Test plan; they verify the event stays unset for an empty-only response and becomes set when playable audio follows an empty prefix.

### CODE REVIEW

File: javascript/src/voice/adapter.runtime.ts [hunk 1]
```diff
@@ -721,8 +721,8 @@ async function drainInner(
   );
   if (first.data.length > 0) {
     onFirstChunk();
+    speakingEvent.set();
```
Comment (langwatch-agent) [inline javascript/src/voice/adapter.runtime.ts:729]: **P1 — A leading empty chunk does not mean the response contains no playable audio.** `drainInner` continues reading after this first chunk, so an input sequence such as `[empty, non-empty, end]` emits audio but never calls `onFirstChunk()` or sets `speakingEvent`. That leaves the speaking/interrupt state inconsistent for a real response. Keep discarding zero-length prefixes until the first playable chunk (or terminal end), then set the event at that point; add a regression test for an empty chunk followed by audio.

**Review by langwatch-agent — CHANGES_REQUESTED** [review]
Requesting changes: 1 P1 correctness issue. See the inline comment.

File: javascript/src/voice/adapter.runtime.ts (hunk 1 above)
Comment (allin2) [reply javascript/src/voice/adapter.runtime.ts:729]: Addressed in `3380f97`.

- `drainInner` now marks the first playable chunk wherever it appears, so a leading empty chunk no longer suppresses `onFirstChunk()` or `speakingEvent.set()`.
- Added the requested regression for `[empty, non-empty, end]`, while retaining the empty-only assertion.
- Verified `interrupt-truncation.test.ts` (12 passed), the Voice suites (436 passed, 1 skipped), and the full JavaScript suite (1036 passed, 4 skipped). Build, lint, typecheck, and CJS/ESM smoke checks also pass.

Could you please re-review?

**Review by langwatch-agent — COMMENTED** [review]
Static review found no actionable voice-runtime or liveness issue in the current fork change. Empty initial chunks no longer signal speech, while the later playable chunk still starts the recorder and speaking event; interruption remains time-bounded elsewhere. I did not execute the untrusted head or its scripts.

**Review by langwatch-agent — COMMENTED** [review]
Static review of external contribution at `3380f97992847641f75449b012c4eb74f9f1da03` (no PR-branch code or dependencies executed). No merge-blocking issue found. The speaking event is now set only after the first playable audio chunk, while an initially empty chunk no longer produces a false speaking transition; the expected empty-first-then-audio and empty-only behavior is covered. CI evidence was considered but not rerun under the external-contribution trust boundary.

**Review by langwatch-agent — COMMENTED** [review]
External static review: no blocking concern found in the current diff. I did not execute branch code, install dependencies, or run contributor-provided scripts. Residual risk: runtime behavior remains covered by the repository CI.

LangWatch-Review: verdict=clean sha=3380f97992847641f75449b012c4eb74f9f1da03 p0=0 p1=0 p2=0 p3=0

**Review by langwatch-agent — COMMENTED** [review]
External static review: no blocking concern found in the current diff. I did not execute branch code, install dependencies, or run contributor-provided scripts. Residual risk: runtime behavior remains covered by the repository CI.

LangWatch-Review: verdict=clean sha=3380f97992847641f75449b012c4eb74f9f1da03 p0=0 p1=0 p2=0 p3=0

**Comment by langwatch-agent** [comment]
## Human Review Brief

> [!CAUTION]
> **This changes when `AgentSpeakingEvent` fires, and interruption logic reads that event.**
>
> Previously the TypeScript voice runtime resolved the event on the first audio chunk, empty or not. An empty first chunk does not prove the agent spoke, so a silent turn could be treated as active speech by interruption logic. The fix waits for the first *playable* chunk, whether that is the initial one or a later one, and leaves the event unset for a response that contains only empty terminal chunks.
>
> Both halves matter and the PR gets both. Gating naively on the first chunk would have missed real audio arriving after an empty prefix, which is the obvious wrong fix. The consequence to think about is the other direction: a turn that produces no playable audio now never resolves the event at all. Anything awaiting it needs to handle that, and the timing of playback start moves later by however long the empty prefix lasts.

| | |
| --- | --- |
| **Mode** | `Fast Skim`. One runtime file plus tests, in the JavaScript SDK. |
| **Issue** | `Closes #569`. Matches. |
| **State** | +34 / -3 across 2 files. CI green, mergeable. Requested reviewer `sergioestebance`. Open since 25 July, so a month. Outside contributor. |
| **Evidence** | Tests cover both the empty-only case and the empty-prefix-then-audio case, which are exactly the two the fix has to separate. Public API preserved. |
| **Where to look** | What awaits `AgentSpeakingEvent`, and what it does when the event never resolves |

<details>
<summary><b>What you are looking at, if the voice subsystem is not your area</b></summary>

Scenario simulates conversations against an agent. The voice subsystem lets the simulated user do that in **audio** rather than text, which means the framework is now responsible for a real-time duplex stream and for the timeline that gets reconstructed from it.

Four facts carry most of the weight.

**It is a deliberate two-language port.** `javascript/src/voice/` mirrors `python/scenario/voice/` file for file, and nearly every TypeScript module names its Python counterpart in its header. Parity is a maintained property here, not a coincidence, so a fix on one side raises the question of the other.

**There is a canonical audio format, and adapters convert at the edge.** Internally everything is an `AudioChunk`: PCM16, mono, 24 kHz. Each provider adapter resamples to and from whatever its API wants (Gemini Live, for instance, takes 16 kHz in and returns 24 kHz out), using plain linear interpolation so the package needs no numeric dependency.

**The runtime, not the adapter, owns the turn.** `adapter.runtime.ts` clears an `AgentSpeakingEvent` (a Promise-paired flag standing in for Python's `asyncio.Event`), drains the agent's audio chunks, and emits the `agent_start_speaking` / `agent_stop_speaking` and `user_start_speaking` / `user_stop_speaking` timeline events. Those events are what the transcript, the judge and the recorded artefacts are built from, so a boundary emitted at the wrong moment is not a cosmetic bug: it moves where the conversation is understood to have happened.

**Interruption is a first-class feature, and it is the hard part.** The simulated user interrupts mid-turn on `interruptProbability`, cutting the agent off. Everything downstream then has to agree about where the turn ended, which is why the runtime has explicit reconciliation that moves the most recent `agent_stop_speaking` when a turn is truncated, and why adapters have to reason about provider content that was already in flight when the cut happened.

One practical note about the tests: the real adapters need live provider credentials, so the suites that run in CI drive fakes, and anything touching a live provider is marked as an integration test and run on demand. A deterministic test here is proving the state machine, not the transport.

</details>

<details>
<summary><b>The one useful check</b></summary>

Find the callers. If anything `await`s `AgentSpeakingEvent` without a timeout, a silent turn now hangs it where before it resolved immediately and wrongly. That is a better failure than the one being fixed and it is still a behaviour change worth being deliberate about.

Playback timing also now starts later, at the first playable chunk rather than the first chunk. If any metric measures time-to-first-audio off that clock, its numbers move.

</details>

> [!NOTE]
> **The event should mean the agent actually spoke. It now does, and the case where it never fires is the one to trace.**

**Comment by langwatch-agent** [comment]
> [!NOTE]
> **Adding the orientation section the brief above should have opened with.** The rest of that brief still stands.

<details open>
<summary><b>What you are looking at, if the voice path is not your directory</b></summary>

Scenario's voice support drives a realtime model over a streaming connection: audio frames go up, audio and events come back, and the library turns that stream into the turn structure the rest of scenario reasons about. A **turn** is one participant speaking until they stop.

The hard part is that the stream does not announce turns cleanly. Frames arrive continuously, silence is not a message, and an interruption means the model was mid-turn when the user started talking again. So the adapters carry a small state machine that decides where one turn ends and the next begins, driven by events like speaking-started and speaking-stopped plus the frames themselves.

That state machine is where nearly every voice defect lives, and it fails in a characteristic way: not a crash, but a turn boundary in the wrong place, which surfaces much later as a judge reading a transcript that does not match what was said. When reading a voice fix, the question is almost always "what does this do to the boundary", not "does this code work".
</details>
